### PR TITLE
add first draft ProviderStore impl using delegated routing client

### PR DESCRIPTION
**Pull request by petar**
- [x] add metrics for wait times (durations for delegated attempts, number of attempted requests, success/failures).
- [x] create a backlog item for more general improved metrics (https://github.com/libp2p/hydra-booster/issues/143)
- [x] add timeout for results from provider stores
- [x] add a mock delegated routing server to be used as a placeholder for testing in production

Addresses https://github.com/libp2p/hydra-booster/issues/140

### CODE REVIEW

**Comment by petar** [comment]
@aschmahmann Take a look: I just updated the hydra code to use the new DHT (with the ProviderStore option). Note that the delegated routing is still not plumbed in. This is just updating the DHT dependence. This seems to have required a bump up on many go mod dependencies and as a result hydra stopped compiling (with various errors). I am going to investigate in detail, but if you have thoughts, let me know.

File: metrics/definitions.go [hunk 1]
```diff
@@ -38,6 +38,12 @@ var (
 	// "failed" (not found any providers on the network)
 	FindProvsDuration  = stats.Float64("find_provs_duration_seconds", "The time it took find provider attempts from the network to succeed or fail because of timeout or completion", stats.UnitSeconds)
 	FindProvsQueueSize = stats.Int64("find_provs_queue_size", "The current size of the queue for finding providers", stats.UnitDimensionless)
+
+	// Augmented with "status" label:
+	// "succeeded" if a response with no error was received from the source.
+	// "failed" if an error was encountered and the request failed.
+	DelegatedFindProvs         = stats.Int64("find_provs_total", "Total delegated find provider attempts that were found locally, or not found locally and succeeded, failed or were discarded", stats.UnitDimensionless)
+	DelegatedFindProvsDuration = stats.Float64("delegated_find_provs_duration_seconds", "The time it took delegated find provider attempts from the network to succeed or fail because of timeout or completion", stats.UnitSeconds)
```
[the remark this answers is not in the document]
Comment (petar) [reply metrics/definitions.go:46]: This was a bug. Fixed it. I had copy-pasted and was using the non-delegated version of this metric. Thanks!

File: providers/combined.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+		go func(backend providers.ProviderStore) {
+			ch <- backend.AddProvider(ctx, key, prov)
+		}(b)
+	}
+	var errs *multierror.Error
+	for range s.backends {
+		if e := <-ch; e != nil {
+			multierror.Append(errs, e)
+		}
+	}
```
[the remark this answers is not in the document]
Comment (petar) [reply providers/combined.go:33]: I think so. I added logging here. There was no logging in hydra at all so far. We should revisit logging eventually and probably establish some conventions, but for now I added a straightforward log of the multierror when some providers fail.

**Comment by BigLep** [comment]
This ideally would use a to-be-created deployment pipeline for Hydras.

**Comment by BigLep** [comment]
For validating this work:
- [ ] Create a dummy server that returns record
- [ ] Do a Hydra deployment that calls the dummy server

**Review by guseggert — APPROVED** [review]

**Comment by petar** [comment]
@aschmahmann This is ready to merge and tested/deployed on alasybil. The Linux test with "-race -tags=openssl" is failing with a networking-looking issue. On MAC, the same test is failing with a different panic issue. Do you know if this test mode is outdated? I can't recall if this test was passing on hydra before.

**Pull request merged**